internal/activity_task_handler_test.go

```diff
@@ -0,0 +1,353 @@
+			WorkflowId: common.StringPtr("wID"),
+			RunId:      common.StringPtr("rID")},
+		ActivityType:                    &s.ActivityType{Name: common.StringPtr("activityWithUserContext")},
+		ActivityId:                      common.StringPtr(uuid.New()),
+		ScheduledTimestamp:              common.Int64Ptr(now.UnixNano()),
+		ScheduledTimestampOfThisAttempt: common.Int64Ptr(now.UnixNano()),
+		ScheduleToCloseTimeoutSeconds:   common.Int32Ptr(1),
+		StartedTimestamp:                common.Int64Ptr(now.UnixNano()),
+		StartToCloseTimeoutSeconds:      common.Int32Ptr(1),
+		WorkflowType: &s.WorkflowType{
+			Name: common.StringPtr("wType"),
+		},
+		WorkflowDomain: common.StringPtr("domain"),
+	}
+	_, err = activityHandler.Execute(tasklist, pats)
+	assert.ErrorIs(t, err, assert.AnError)
+}
+
+func TestActivityTaskHandler_Execute_with_auto_heartbeat(t *testing.T) {
+	logger := zaptest.NewLogger(t)
+
+	now := time.Now()
+
+	clock := clockwork.NewFakeClock()
+
+	// inline activity using value passing through user context.
+	activityWithUserContext := func(ctx context.Context) error {
+		// heartbeat once
+		clock.Advance(time.Second / 2)
+		return nil
+	}
+	registry := newRegistry()
+	err := registry.registerActivityFunction(activityWithUserContext, RegisterActivityOptions{
+		Name:                "activityWithUserContext",
+		EnableAutoHeartbeat: true,
+	})
+	require.NoError(t, err)
+
+	mockCtrl := gomock.NewController(t)
+	mockService := workflowservicetest.NewMockClient(mockCtrl)
+	wep := workerExecutionParameters{
+		WorkerOptions: WorkerOptions{
+			Logger:        logger,
+			DataConverter: getDefaultDataConverter(),
+			Tracer:        opentracing.NoopTracer{},
+		},
+	}
+	ensureRequiredParams(&wep)
+	activityHandler := newActivityTaskHandler(mockService, wep, registry)
+	activityHandlerImpl := activityHandler.(*activityTaskHandlerImpl)
+	activityHandlerImpl.clock = clock
+
+	pats := &s.PollForActivityTaskResponse{
+		TaskToken: []byte("token"),
+		WorkflowExecution: &s.WorkflowExecution{
+			WorkflowId: common.StringPtr("wID"),
+			RunId:      common.StringPtr("rID")},
+		ActivityType:                    &s.ActivityType{Name: common.StringPtr("activityWithUserContext")},
+		ActivityId:                      common.StringPtr(uuid.New()),
+		ScheduledTimestamp:              common.Int64Ptr(now.UnixNano()),
+		ScheduledTimestampOfThisAttempt: common.Int64Ptr(now.UnixNano()),
+		ScheduleToCloseTimeoutSeconds:   common.Int32Ptr(1),
+		StartedTimestamp:                common.Int64Ptr(now.UnixNano()),
+		StartToCloseTimeoutSeconds:      common.Int32Ptr(1),
+		HeartbeatTimeoutSeconds:         common.Int32Ptr(1),
+		WorkflowType: &s.WorkflowType{
+			Name: common.StringPtr("wType"),
+		},
+		WorkflowDomain: common.StringPtr("domain"),
+	}
+	res, err := activityHandler.Execute(tasklist, pats)
+	require.NoError(t, err)
+	_, ok := res.(*s.RespondActivityTaskCompletedRequest)
+	require.True(t, ok, "response is not of type *s.RespondActivityTaskCompletedRequest")
```

Comment:
nit: include the actual type of `res` in the message

Reply:
Good point. Added.